--- original/quickjs/qjsc.c	2022-11-18 18:50:21.000000000 +0800
+++ upstream/quickjs/qjsc.c	2022-11-27 09:15:32.000000000 +0800
@@ -27,14 +27,17 @@
 #include <inttypes.h>
 #include <string.h>
 #include <assert.h>
-#include <unistd.h>
 #include <errno.h>
 #if !defined(_WIN32)
 #include <sys/wait.h>
+  #include <unistd.h>
+#else
+  #include "win/getopt.h"
 #endif
 
 #include "cutils.h"
 #include "quickjs-libc.h"
+#include "quickjs-version.h"
 
 typedef struct {
     char *name;
@@ -341,7 +344,7 @@
 
 void help(void)
 {
-    printf("QuickJS Compiler version " CONFIG_VERSION "\n"
+    printf("QuickJS Compiler version " QUICKJS_VERSION "\n"
            "usage: " PROG_NAME " [options] [files]\n"
            "\n"
            "options are:\n"
