--- original/quickjs/quickjs-opcode.h	2022-11-18 18:50:21.000000000 +0800
+++ upstream/quickjs/quickjs-opcode.h	2022-11-27 09:15:32.000000000 +0800
@@ -260,6 +260,9 @@
 DEF(      mul_pow10, 1, 2, 1, none)
 DEF(       math_mod, 1, 2, 1, none)
 #endif
+#ifdef CONFIG_DEBUGGER
+DEF(line_num, 5, 0, 0, u32) /* emitted in phase 1 and kept if running under debugger */
+#endif
 /* must be the last non short and non temporary opcode */
 DEF(            nop, 1, 0, 0, none) 
 
@@ -283,7 +286,9 @@
 
 def( set_class_name, 5, 1, 1, u32) /* emitted in phase 1, removed in phase 2 */
     
+#ifndef CONFIG_DEBUGGER
 def(       line_num, 5, 0, 0, u32) /* emitted in phase 1, removed in phase 3 */
+#endif
 
 #if SHORT_OPCODES
 DEF(    push_minus1, 1, 0, 1, none_int)
